<a href="https://colab.research.google.com/github/gracewindheim/gmwDataSci/blob/master/q1hw8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Question 1

Use Keras and Tensor Flow on colab to fit a prediction model for the mnist data with multiple hidden layers. The first has 128 nodes, the second 64 and the 3rd 32 then an output layer of size 10. Have 20% dropout for each hidden layer. Compare the test set accuracy to a model with only one hidden layer with 32 nodes and then an output layer of size 10. Have 20% dropout for the hidden layer.

In [0]:
import tensorflow as tf
import matplotlib.pyplot as plt
import numpy as np

mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0


11493376/11490434 [==============================] - 0s 0us/step


## Model version 1:

*   hidden layer of 128 nodes
*   hidden layer of 64 nodes
* hidden layer of 32 nodes
* output layer size 10

20% dropout each hidden layer



In [0]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(64, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(32, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10)
])

model.summary()

loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_2 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_8 (Dense)              (None, 128)               100480    
_________________________________________________________________
dropout_6 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_9 (Dense)              (None, 64)                8256      
_________________________________________________________________
dropout_7 (Dropout)          (None, 64)                0         
_________________________________________________________________
dense_10 (Dense)             (None, 32)                2080      
_________________________________________________________________
dropout_8 (Dropout)          (None, 32)               

In [0]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 0s - loss: 0.0891 - accuracy: 0.9746


[0.08906542509794235, 0.9746000170707703]

## Model version 2:

* hidden layer of 32 nodes
* output layer size 10

20% dropout for the hidden layer

In [0]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(32, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10)
])

model.summary()

loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_3 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_12 (Dense)             (None, 32)                25120     
_________________________________________________________________
dropout_9 (Dropout)          (None, 32)                0         
_________________________________________________________________
dense_13 (Dense)             (None, 10)                330       
Total params: 25,450
Trainable params: 25,450
Non-trainable params: 0
_________________________________________________________________
Epoch 1/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.4800 - accuracy: 0.8566
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2813 - accuracy: 0.9164
Epoch 3/5
1875/1875 [=========================

In [0]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 0s - loss: 0.1417 - accuracy: 0.9586


[0.14174287021160126, 0.9585999846458435]

## Comparison

Accuracy for model version 1 (3 hidden layers) = 97.5%

Accuracy for model version 2 (1 hidden layers) = 95.9%

The first model with more hidden layers was slightly more accurate. This could be due to the ability of the model to approximate complexity of the relationship between input and output variables due to more layers and node connections.